Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift

```diff
@@ -570,8 +615,200 @@ extension MessageHistoryEndpointTests {
+      .messageCounts(channels: testSingleChannel, timetoken: 0) { result in
+        switch result {
+        case .success:
+          XCTFail("This should fail")
+        case let .failure(error):
+          XCTAssertEqual(error.pubNubError, PubNubError(.messageHistoryNotEnabled))
+        }
+        expectation.fulfill()
+      }
+
+    wait(for: [expectation], timeout: 1.0)
+  }
+}
+
+// MARK: - MessageHistory ResponseDecoder
+
+extension HistoryRouterTests {
+
+  func testFetchV2_decode_jsonDataDecodingFailure() {
+    let decoder = MessageHistoryResponseDecoder()
+    let router = HistoryRouter(.fetchV2(channel: testChannel, max: nil, start: nil, end: nil, includeMeta: false), configuration: config)
+    let response = EndpointResponse(router: router,
+                                    request: .init(url: URL(string: "http://example.com")!),
+                                    response: .init(),
+                                    payload: Data())
+
+    switch decoder.decode(response: response) {
+    case .success:
+      XCTFail("Decoder should not succeed")
+    case .failure(let error):
+      XCTAssertEqual(error.pubNubError?.reason, .jsonDataDecodingFailure)
+    }
+  }
+
+  func testFetchV2_decode_malformedResponse() {
+    guard let data = AnyJSON([]).jsonData, let url = URL(string: "http://example.com") else {
+      return XCTFail("Could not create JSON data")
+    }
+    let decoder = MessageHistoryResponseDecoder()
+    let router = HistoryRouter(.fetchV2(channel: testChannel, max: nil, start: nil, end: nil, includeMeta: false), configuration: config)
+    let response = EndpointResponse(router: router,
+                                    request: .init(url: url),
+                                    response: .init(),
+                                    payload: data)
+
+    guard let result = try? decoder.decodeMessageHistoryV2(response: response) else {
+      return XCTFail("Deocder threw an exception")
+    }
+    switch result {
+    case .success:
+      XCTFail("Decoder should not succeed")
+    case .failure(let error):
+      XCTAssertEqual(error.pubNubError?.reason, .malformedResponseBody)
+    }
+  }
 
-extension MessageHistoryEndpointTests {}
+  func testMessageHistoryResponse_Deocde_missingChannels() {
+    let data: [String: JSONCodableScalar] = ["status": 0, "error": false, "error_message": "Test Message"]
 
+    guard let json = data.mapValues({ $0.codableValue }).jsonData else {
+      return XCTFail("Could not convert object to data")
+    }
+
+    XCTAssertEqual(try? JSONDecoder().decode(MessageHistoryResponse.self, from: json).channels.isEmpty, true)
+  }
+
+  func testMessageHistoryChannelPayload_Decode_InvalidTimetokenStrings() {
+    let data: [Any] = []
+
+    guard let json = AnyJSON(data).jsonData else {
+      return XCTFail("Could not convert object to data")
+    }
+
+    let payload = try! JSONDecoder().decode(MessageHistoryChannelPayload.self, from: json)
+
+
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Limit vertical whitespace to a single empty line. Currently 2.](https://app.codacy.com/app/PubNub/swift/pullRequest?prid=4513555)